crypto/s2n_certificate.c

```diff
@@ -605,3 +611,113 @@ int s2n_get_cert_der(const struct s2n_cert *cert, const uint8_t **out_cert_der,
 
     return S2N_SUCCESS;
 }
+
+static int s2n_asn1_string_free(ASN1_STRING** data)
+{
+    if (*data != NULL) {
+        ASN1_STRING_free(*data);
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_get_utf8_string_from_extension_data(const uint8_t *extension_data, uint32_t extension_len, uint8_t **out_data, uint32_t *out_len)
+{
+    POSIX_ENSURE_REF(extension_data);
+    POSIX_ENSURE_GT(extension_len, 0);
+    POSIX_ENSURE_REF(out_data);
+    POSIX_ENSURE_REF(out_len);
+
+    DEFER_CLEANUP(ASN1_STRING *asn1_str = NULL, s2n_asn1_string_free);
+    const uint8_t *asn1_str_data = extension_data;
```

Comment:
why is this needed?

Reply:
This is needed due to pointer manipulation that OpenSSL does. I have a comment for it on the next line as it relates to `d2i_ASN1_UTF8STRING`.  Here, d2i_ASN1_UTF8STRING increments *der_in to the byte following the parsed data. For all `d2i_TYPE` APIs this is a known bug and using a temporary variable is mandatory for these APIs. Refer Warning section here for more information: https://www.openssl.org/docs/man1.1.0/man3/d2i_X509.html

Note: d2i_TYPE() attempts to decode len bytes at *ppin. If successful a pointer to the TYPE structure is returned and *ppin is incremented to the byte following the parsed data. If a is not NULL then a pointer to the returned structure is also written to *a. If an error occurred then NULL is returned.